### DISCUSSION TITLE
How Can I Select a HTML element with Blazor

### DISCUSSION THREAD

**Discussion by AM-Nateghi** [dotnet/AspNetCore.Docs#33076] [Q&A]
Hello!
How can I select an HTML element/tag with Blazor attributes (without using @ref)? I didn't find any feature in it, which could help me.
I want something like javascript selector methods or jquery selector.

But still, for now, I'm using JQuery by JSInterop. However, I still have a lot of problems!
After running my Blazor app for the first time, the jQuery methods and JavaScript code are working. But after changing the page, the old javascript files remain on my web page and break my app!
To make matters worse, new JavaScript files not working means that the new JavaScript files are loaded and available, but not executed.

can you help me?

**Comment by Rick-Anderson** [comment]
Have you tried asking CoPilot? Copy/paste of your questions results in:
Selecting HTML Elements in Blazor without @ref
While Blazor doesn’t have built-in support for selecting HTML elements like jQuery, you can still achieve this using JavaScript interop. Here’s an example of how you can use JavaScript to select elements:

Define a JavaScript function to select elements:
```JavaScript

function selectElement(selector) {
    return document.querySelector(selector);
}
```
//Call this function from Blazor using JSInterop:

```
@inject IJSRuntime JSRuntime

@code {
    private async Task SelectElementAsync(string selector)
    {
        await JSRuntime.InvokeVoidAsync("selectElement", selector);
    }
}

```

Handling JavaScript Files on Page Change
The issue with JavaScript files not being executed after changing pages is common in Blazor applications. Here are a few tips to resolve this:

Ensure JavaScript files are loaded correctly:
[Move your script tags to the index.html or _Host.cshtml file to ensure they are loaded once and remain available throughout the app](https://www.talkingdotnet.com/using-jquery-with-asp-net-core-blazor/)[1](https://www.talkingdotnet.com/using-jquery-with-asp-net-core-blazor/).
Reinitialize JavaScript on page change:
Use the OnAfterRenderAsync method to reinitialize your JavaScript functions after each render:


```
@inject IJSRuntime JSRuntime

protected override async Task OnAfterRenderAsync(bool firstRender)
{
    if (firstRender)
    {
        await JSRuntime.InvokeVoidAsync("initializeScripts");
    }
}
```

Avoid modifying the DOM directly:
Blazor maintains its own representation of the DOM. [Direct modifications using JavaScript can lead to inconsistencies](https://www.talkingdotnet.com/using-jquery-with-asp-net-core-blazor/)[2](https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/?view=aspnetcore-8.0). Always use Blazor’s built-in methods or JSInterop to interact with the DOM.

@guardrex  will tell me if this is incorrecto
